contracts/Commitment/StateMachine.sol

```diff
@@ -0,0 +1,124 @@
+pragma solidity 0.4.15;
+
+import './MStateMachine.sol';
+
+
+//
+// Before --> Whitelist --> Public --> Finished
+//
+contract StateMachine is MStateMachine {
+
+    ////////////////////////
+    // Mutable state
+    ////////////////////////
+
+    State internal _state;
+
+    ////////////////////////
+    // Events
+    ////////////////////////
+
+    event LogStateTransition(
+        State oldState,
+        State newState
+    );
+
+    ////////////////////////
+    // Modifiers
+    ////////////////////////
+
+    modifier onlyState(State state) {
+        require(_state == state);
+        _;
+    }
+
+    modifier onlyStates(State state0, State state1) {
+        require(_state == state0 || _state == state1);
+        _;
+    }
+
+    modifier onlyStates3(State state0, State state1, State state2) {
+        require(_state == state0 || _state == state1 || _state == state2);
+        _;
+    }
+
+    /// @dev Multiple states can be handled by adding more modifiers.
+    modifier notInState(State state) {
+        require(_state != state);
+        _;
+    }
+
+    modifier transitionsTo(State newState) {
+        _;
+        transitionTo(newState);
+    }
+
+    ////////////////////////
+    // Constructor
+    ////////////////////////
+
+    function StateMachine() {
+        _state = State.Before;
+    }
+
+    ////////////////////////
+    // Public functions
+    ////////////////////////
+
+    function state()
+        public
+        constant
+        returns (State)
+    {
+        return _state;
+    }
+
+    ////////////////////////
+    // Internal functions
+    ////////////////////////
+
+    function transitionTo(State newState)
+        internal
+        returns (State oldState)
+    {
+        oldState = _state;
+        if (oldState == newState) {
+            return;
+        }
+        require(validTransition(oldState, newState));
```

Comment:
another case not to return on cycling: it may be prohibited by `validTransition` function. let it execute

Reply:
I prefer the current semantics of silently ignoring.

Reply:
heh OK, my thinking was: better reject in testing than silently push to production. it is academic discussion. this state machine is too simple to have this case even by mistake